bullet-featherstone: Use a single contact manifold for each convex decomposed mesh collision

iche033 opened the pull request:
# 🦟 Bug fix

## Summary

Contact point optimization for convex decomposed mesh collisions in bullet-featherstone plugin. This PR reduces the number of contact points generated for convex decomposed mesh collision by using a single contact manifold.

Previously when a mesh collision is decomposed into multiple convex hull shape, each convex hull shape has a corresponding contact manifold which can generate up to 4 contact points (4 is the default limit in bullet). This slows down sim especially when querying contact points (via `GetContactsFromLastStep`). This PR overrides the `btCollisionDispatcher` class in order to update the contact manifolds stored in bullet. It post processes the contact points, generates a single contact manifold for each collision associated with convex hull shapes, and clears the exiting contact manifolds. 

Note: it would be more efficient if we can do this inside bullet to avoid generating all the contact manifolds in the first place.

Before: many points can be generated from one convex decomposed mesh collision.
<img width="286" alt="bullet_mesh_contacts" src="https://github.com/gazebosim/gz-physics/assets/4000684/366244a3-6211-4199-a6fe-6f061c6b4cb3">

After: limit contact points to <= 4 
<img width="209" alt="bullet_mesh_contacts_single_manifold" src="https://github.com/gazebosim/gz-physics/assets/4000684/6632a20f-81fa-40c4-827c-bb8f79956ee8">


Here're remotery graphs showing before and after the fix for a scene with multiple convex decomposed meshes. Note objects were auto-deactivated when these graphs were captured.

Before: >200 contact points
<img width="843" alt="bullet_deactivated_profile_cropped" src="https://github.com/gazebosim/gz-physics/assets/4000684/88193af6-8add-47de-bb95-ee7ef88734b9">

After: 20 contact points
<img width="944" alt="bullet_deactivated_with_fix_profile_cropped" src="https://github.com/gazebosim/gz-physics/assets/4000684/0b7e1002-cff1-4503-b593-963fb75886c7">

Notable speedups are:
* `PhysicsPrivate::Step`: 0.153ms -> 0.097ms
* `PhysicsPrivate::UpdateCollision`:  0.528ms -> 0.135ms

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

iche033 committed:
Use 1 contact manifold for each convex decomposed mesh collision

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
fix

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
style

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
remove unused features in test

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
use 2 separate sets to keep track of manifolds, update logic to find collision in get contacts from last step

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
Merge branch 'gz-physics7' into bullet_convex_manifold

iche033 committed:
undo some changes

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
style

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
Merge branch 'gz-physics7' into bullet_convex_manifold

azeey commented:
Looks like there are several test failures in windows and I don't see any in the stable branch (https://build.osrfoundation.org/view/gz-harmonic/job/gz_physics-7-win/)

iche033 committed:
ifdef for windows

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
undo changes in simulation features

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
test with GzCollisionDispatcher

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
testing windows

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
more windows testing

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
more testing on win

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
rename var

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
revert ba6f4d02bdd8a62473f6881b0fe9581dba448024

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
return early

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
debugging

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
debugging

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
add more debugging

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
check nulls

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
more debugging

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
test static cast

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
test static cast

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
enable gz collision dispatcher

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
cleanup

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 committed:
more cleanup

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 commented:
> Looks like there are several test failures in windows and I don't see any in the stable branch (https://build.osrfoundation.org/view/gz-harmonic/job/gz_physics-7-win/)

Fixed and this PR is ready for review again. The test failures were due to  RTTI access violation when doing a `dynamic_cast` from `btCollisionShape` to `btCompoundShape`. I saw a similar [issue](https://github.com/dartsim/dart/issues/1522) reported in dartsim and [fixed](https://github.com/dartsim/dart/pull/1541) by adding pragmas in base classes but that's not viable here so I switched to `static_cast`.

iche033 committed:
Merge branch 'gz-physics7' into bullet_convex_manifold

iche033 committed:
Merge branch 'gz-physics7' into bullet_convex_manifold

iche033 committed:
Merge branch 'gz-physics7' into bullet_convex_manifold

azeey commented:
@azeey to review.

iche033 committed:
Merge branch 'gz-physics7' into bullet_convex_manifold

azeey commented on bullet-featherstone/src/Base.cc:
```diff
@@ -26,14 +26,213 @@ namespace gz {
+        continue;
+      }
+
+      btPersistentManifold* colManifold =
```
When we remove manifolds, we don't do anything with the pointer in `colPairManifolds`, so we it might not be a nullptr even if the manifold has been removed. I suggest checking if it's in `customManifolds` instead of just a `nullptr` check.

iche033 replied:
I updated the code to remove the pointers in `colPairManifolds` when manifolds are removed. So the `colPairManifolds` map should always be up-to-date now. [15d4ede](https://github.com/gazebosim/gz-physics/pull/664/commits/15d4ede77239f9a1b35971a13a9c3fe4ccaa1bca)

azeey commented on bullet-featherstone/src/Base.cc:
```diff
@@ -26,14 +26,213 @@ namespace gz {
+  auto manifoldIt = this->manifoldsToKeep.find(_manifold);
+  if (manifoldIt != this->manifoldsToKeep.end())
+    this->manifoldsToKeep.erase(manifoldIt);
+
```
Do we need to also remove it from `customManifolds`?

iche033 replied:
I think this is only be called for contact manifolds owned by bullet, and not the custom ones created by gz-physics. I tested manually by removing models with convex decomposed collisions and did not notice that custom manifolds were passed to this function.

azeey commented on bullet-featherstone/src/Base.hh:
```diff
@@ -551,6 +606,10 @@ class Base : public Implements3d<FeatureList<Feature>>
+        GzCollisionDispatcher *dispatcher =
+            dynamic_cast<GzCollisionDispatcher *>(
+            world->world->getDispatcher());
+        dispatcher->RemoveManifoldByCollisionObject(link->collider.get());
```
Should we remove the object from the `colPairManifolds` map too? Otherwise, that map will grow indefinitely, which might become a problem if objects are added and removed frequently.

iche033 replied:
yes removed them from `colPairManifolds`. [15d4ede](https://github.com/gazebosim/gz-physics/pull/664/commits/15d4ede77239f9a1b35971a13a9c3fe4ccaa1bca)

azeey commented on bullet-featherstone/src/Base.cc:
```diff
@@ -26,14 +26,213 @@ namespace gz {
+    if (this->customManifolds.find(contactManifold) !=
+        this->customManifolds.end())
+    {
+      contactManifold->refreshContactPoints(ob0->getWorldTransform(),
```
Is this necessary? Do you know that Bullet itself won't be doing this?

iche033 replied:
I remember I had to keep this call for the multi body link transform to be updated. As a test, I dropped a model with convex decomposed collisions on the ground -> this worked fine, but as soon as I moved it to a different pose, it's transforms are not updated.

azeey commented on bullet-featherstone/src/Base.cc:
```diff
@@ -26,14 +26,213 @@ namespace gz {
+      // collisions alongside of other collisions. See following example:
+      // parentLink -> boxShape0
+      //            -> boxShape1
+      //            -> comopundShape -> convexShape0
```
```suggestion
      //            -> compoundShape -> convexShape0
```

azeey commented on bullet-featherstone/src/Base.cc:
```diff
@@ -26,14 +26,213 @@ namespace gz {
+      //            -> comopundShape -> convexShape0
+      //                             -> convexShape1
+      // A _childIndex of 1 is ambiguous as it could refer to either
+      // boxShape1 or convexShape1
```
Could we use the `getUserIndex` or `getUserPointer` to help with this somehow?

azeey reviewed (commented):
Well, this was not a walk in the park :), but I think I finally understand how things are working. I don't see any major problems, but it would be good to test what happens when objects are removed. Another point is about the ambiguity with the collision index. I'm wondering if we can find a way around that. No worries if you don't have time to tackle that right now though.

iche033 committed:
update colPairManifolds on removal, add test

Signed-off-by: Ian Chen <ichen@openrobotics.org>

iche033 commented on bullet-featherstone/src/Base.hh:
```diff
@@ -53,6 +53,61 @@ namespace gz {
+class GzCollisionDispatcher : public btCollisionDispatcher
+{
+  /// \brief Constructor
+  public: GzCollisionDispatcher(
```
[15d4ede](https://github.com/gazebosim/gz-physics/pull/664/commits/15d4ede77239f9a1b35971a13a9c3fe4ccaa1bca)

iche033 commented on bullet-featherstone/src/Base.cc:
```diff
@@ -26,14 +26,213 @@ namespace gz {
+}
+
+/////////////////////////////////////////////////
+void GzCollisionDispatcher::RemoveManifoldByCollisionObject(
```
updated `MeshContact` test in `collisions.cc` test to remove the model with convex decomposed mesh collision. It should trigger this call

iche033 commented:
I addressed most of the comments but not all. I still need to figure out how to resolve the issue with ambiguous collision index and then add a test.
